### ISSUE TITLE
New package for LC-MS/metabolomics feature grouping?

### ISSUE THREAD

**Issue by jorainer** [rformassspectrometry/MsFeatures#1]
We've implemented some functionality that allows to group LC-MS *features* (i.e. defined by their m/z and retention time range) into *feature group* where ideally each feature group collects all features that come from the same original compound (i.e. are adducts or isotopes ... of that). The question now is what should be the home of this functionality. Currently 3 main grouping functions are available:

- group features based on similar retention time
- group features based on high correlation of feature abundances across samples
- group features based on similar peak shape of the EIC

The LC-MS feature grouping functions above re-use functionality from `xcms` (especially for the correlation based on peak shape of the EIC).

The general workflow would be the following: 
`xcms` (peak picking, feature definitions) -> ??? (initial feature grouping based on properties of the LC-MS) -> `Features` (optional further feature grouping, independent of LC-MS).

The question now is where should be put the functionality. Possible options are:

1) put it into `xcms`
2) put it into `Features`
3) put it into a new package `MetaboFeatures` or `xcmsFeatures`.

Would be nice to get some feedback @lgatto @sneumann @stanstrup @michaelwitting @sgibb

**Comment by cbroeckl** [comment]
quite some time ago after developing ramclustR i was interested in moving some of the approaches from it into CAMERA.  This package, if it is reimplementing CAMERA for feature grouping, could be an opportunity to reinvest in that effort.   @stanstrup @jorainer  - i would like to work towards this if you think it suitable.   There are a few differences (by my understanding) between the ramclustR approach and CAMERA.   

1.  Feature Grouping.   If my memory serves me (it certainly may be failing me...) CAMERA is pretty sequential in how it is assigning ions to groups.   i should probably revisit the source code/approach before making any conclusions, but i recall contrasting this with the ramclustR approach which generates a complete similarity network of all features to each other before pruning the full network into groups using the dynamicTreeCut package.  
2. Feature annotation within groups.   CAMERA assumes that there are multiple molecules in each groups, ramclustR assumes there is one molecule in each group.   The flipside of this:  CAMERA assumes that unexplained ions (those which cannot be annotated by the given rules) are second (third, fourth...) molecules, ramclustR assumes they derive from the same compound in some unexplained manner.   

I am sure this is oversimplifying at least a bit, but this package will enable us to focus on the common goals of the two packages, and return a common data structure which better integrates with the rest of the R mass spec package family, which i would be excited to try to help with.

**Comment by jorainer** [comment]
This sounds really great Corey! I would love to integrate ramclustR with the `MsFeatures` package. The idea of `MsFeatures` is pretty simple, it takes an input object (`XCMSnExp` or `SummarizedExperiment`) and groups the features in it defining a `character` vector which represents the grouping (length of the vector is the same as there are features), grouped features will have the same *feature group ID*. 

Limitations:
- assignment is binary, i.e. a feature can only be part of one group.

Advantages:
- very easy to add new grouping approaches
- potential to combine (sequentially) grouping approaches

If we join forces I think we should be ablel to add that functionality preatty easily... I hope. Can you point me to some code/documentation I could start digging into ramclustR to better understand it?

**Comment by cbroeckl** [comment]
https://github.com/cbroeckl/RAMClustR/blob/master/R/rc.ramclustr.R

ramclustR adheres to your description - assignment is binary.   the premise is to calculate the similarity matrix for all features, based on retention time similarity and quantitative similarity over the sample set (pearson's r).  I didn't use peak shape initially, mostly due to a lack of skill in extracting that many EICs, but also because i was also ramclustR for grouping DIA fragment ions, derived from MSe data, which would have required a bit more code to properly adjust to align with the MS1.   to manage memory, the full similarity matrix is calculated in (generally) 2000 feature chunks and the data is stored in an ff object.  i explored sparse matrix implementations in the past, but that may be worth revisiting. 

After you have a full similarity matrix, you cluster using HCA.  memory is again a bit of a concern here, and it would good to look into the latest and greatest for reducing memory burden.  

the dynamicTreeCut package is used to then cut the full HCA dendrogram into clusters, ideally representing a compound per cluster.   The output of this is a numeric vector of length = n.features.   I generally then sign a character name string to each compound as well:  cluster #1 becomes 'C0001' etc.  

happy to help, of course, just let me know when you plan to start working in this area.  I don't know if my code is intelligible or not ;-).

**Comment by jorainer** [comment]
Thanks for the update! And sorry for my ignorance, but then it sounds like ramclustR groups based *only* on the MS1 properties retention time and quantified feature value and not considering also MS2 spectra? Is that correct?

**Comment by cbroeckl** [comment]
Correct.  the algorithm does not use MS/MS data at all to assign clusters.  The only properties are retention time and quantified feature values.  

That said, it was originally developed with MSe in mind, and we can perform centWave signal detection on the MSe data, where MSe is full mass range MS/MS - i.e. the least selective form of DIA imaginable.  Since MSe samples all precursors as frequently as we have MS1 level scans, we can use CentWave on MSe (MS2) chromatograms as well.  If that has been done, RAMClustR can also cluster MS2 fragment ions to generate reconstructed MS2 spectra from MSe data.  

In the MSe case, XCMS is performed on MS1 and MS2 data for each injection, the full xcms set is aligned and ramclustr splits MS1 from MS2 data.  MS1 data is used for quantitative signal intensity, the MS2 data are used for annotation.   MSe data is not required for clustering, but if you have MSe data processed with the MS1 data, we can use both to improve annotation.   

It is an open question as to whether we should build that part (MSe, AIF) into this package.
